src/client/src/producer.rs

```diff
@@ -116,22 +135,26 @@ async fn send_record_raw<F: SerialFrame>(
     let mut topic_request = DefaultTopicRequest::default();
     let mut partition_request = DefaultPartitionRequest::default();
 
-    debug!(
-        "send record {} bytes to: replica: {}, {}",
-        value.len(),
-        replica,
-        leader
-    );
-
-    let mut record_msg = DefaultRecord {
-        value: DefaultAsyncBuffer::new(value.to_owned()),
-        ..Default::default()
-    };
-    if let Some(key) = key {
-        record_msg.key = Some(key.into());
-    }
+    debug!("Putting together batch with {} records", records.len());
+
     let mut batch = DefaultBatch::default();
-    batch.add_record(record_msg);
+    for (key, value) in records {
```

Comment:
can you `map` iterator rather than `for..in`

Reply:
I just took a look and tried to do this, but it turned out to be trickier than it looks. Usually we can just use iterators to map and collect things, but we cannot collect directly into a DefaultBatch and we also cannot directly assign the internal records Vec, we need to use `add_record` for each individual record. So right now the for loop seems like the best option. Perhaps in the future we could add something to the API of DefaultBatch to make this nicer.